R/dummy_extract.R

```diff
@@ -0,0 +1,280 @@
+#'   for splitting. [strsplit()] is used to perform the split. `sep` takes
+#'   priority if `pattern` is also specified.
+#' @param pattern Character vector containing a regular expression used
+#'   for extraction. [gregexpr()] and [regmatches()] are used to perform
+#'   pattern extraction using `perl = TRUE`.
+#' @template step-return
+#' @family dummy variable and encoding steps
+#' @seealso [dummy_names()]
+#' @export
+#' @details `step_dummy_extract()` will create a set of integer dummy
+#'  variables from a character variable by extract individual strings
+#'  by either splitting or extracting then counting those to create
+#'  count variables.
+#'
+#'  Note that `threshold` works in a very specific way for this step.
+#'  While it is possible for one label to be present multiple times in
+#'  the same row, it will only be counted once when calculating the
+#'  occurrences and frequencies.
+#'
+#' @template dummy-naming
+#'
+#' @details
+#'
+#'  # Tidying
+#'
+#'  When you [`tidy()`][tidy.recipe()] this step, a tibble with columns
+#'  `terms` (the selectors or original variables selected) and `columns`
+#'  (the list of corresponding columns) is returned. The `columns` is
+#'  is ordered according the frequency in the training data set.
+#'
+#' @examples
+#' library(modeldata)
+#' data(tate_text)
+#'
+#' dummies <- recipe(~ artist + medium, data = tate_text) %>%
+#'     step_dummy_extract(artist, medium, sep = ", ") %>%
+#'     prep()
+#'
+#' dummy_data <- bake(dummies, new_data = NULL)
+#'
+#' dummy_data %>%
+#'     select(starts_with("medium")) %>%
+#'     names()
+#'
+#' # More detailed splitting
+#' dummies_specific <- recipe(~ medium, data = tate_text) %>%
+#'     step_dummy_extract(medium, sep = "(, )|( and )|( on )") %>%
+#'     prep()
+#'
+#' dummy_data_specific <- bake(dummies_specific, new_data = NULL)
+#'
+#' dummy_data_specific %>%
+#'     select(starts_with("medium")) %>%
+#'     names()
+#'
+#' tidy(dummies, number = 1)
+#' tidy(dummies_specific, number = 1)
+#'
+#' # pattern argument can be useful to extract harder patterns
+#' color_examples <- tibble(
+#'   colors = c("['red', 'blue']",
+#'              "['red', 'blue', 'white']",
+#'              "['blue', 'blue', 'blue']")
+#' )
+#'
+#' dummies_color <- recipe(~ colors, data = color_examples) %>%
+#'   step_dummy_extract(colors, pattern = "(?<=')[^',]+(?=')") %>%
+#'   prep()
+#'
+#' dommies_data_color <- dummies_color %>%
+#'   bake(new_data = NULL)
+#'
+#' dommies_data_color
+step_dummy_extract <-
+  function(recipe,
+           ...,
+           role = "predictor",
+           trained = FALSE,
+           sep = NULL,
+           pattern = NULL,
+           threshold = 0.0,
+           other = "other",
+           naming = dummy_names,
+           levels = NULL,
+           keep_original_cols = FALSE,
+           skip = FALSE,
+           id = rand_id("dummy_extract")) {
+
+    if (!is_tune(threshold) & !is_varying(threshold)) {
```

Comment:
The help for `threshold` says: 

> A numeric value between 0 and 1, *or an integer greater or equal to one*. If less than one, then factor levels with a rate of occurrence in the training set below threshold will be pooled to other. *If greater or equal to one, then this value is treated as a frequency and factor levels that occur less than threshold times will be pooled to other.*

I changed the threshold checking code to be similar to `step_other()` but the filtering (and testing) does not account for threshold >= 1. 

`step_other()`does _not_ allow a threshold of zero (which would be not othering). This should probably be changed too.

Reply:
threshold can now be >= 1, with tests here:

https://github.com/tidymodels/recipes/pull/901/files#diff-00368cb3590d392abc25294e248ce57469605f0c0eabaffc90b2374ff3309943R135-R205